hs/src/Reach/Connector/ALGO.hs

```diff
@@ -1938,34 +1923,40 @@ cm km = \case
     when recordNew $
       addNewTok $ DLA_Var dv
     sallocVarLet (DLVarLet (Just vc) dv) sm (ce de) km
-  DL_ArrayMap at ansv aa lv iv (DLBlock _ _ body ra) -> do
+  DL_ArrayMap at ansv as xs iv (DLBlock _ _ body ra) -> do
     let anssz = typeSizeOf $ argTypeOf $ DLA_Var ansv
-    let (_, xlen) = argArrTypeLen aa
+    let xlen = arraysLength as
     let rt = argTypeOf ra
     check_concat_len anssz
     salloc_ (textyv ansv) $ \store_ans load_ans -> do
       cbs ""
       store_ans
       cfor xlen $ \load_idx -> do
         load_ans
-        doArrayRef at aa True $ Right load_idx
-        sallocLet lv (return ()) $
-          store_let iv True load_idx $ do
-            cp (ca ra >> ctobs rt) body
+        let finalK = cp (ca ra >> ctobs rt) body
+        let bodyF (x, a) k () = do
+             doArrayRef at a True $ Right load_idx
+             sallocLet x (return ()) $
+               store_let iv True load_idx $
+               k
+        (F.foldrM bodyF finalK $ zip xs as) ()
         op "concat"
         store_ans
       store_let ansv True load_ans km
-  DL_ArrayReduce at ansv aa za av lv iv (DLBlock _ _ body ra) -> do
-    let (_, xlen) = argArrTypeLen aa
+  DL_ArrayReduce at ansv as za av xs iv (DLBlock _ _ body ra) -> do
+    let xlen = arraysLength as
     salloc_ (textyv ansv) $ \store_ans load_ans -> do
       ca za
       store_ans
       store_let av True load_ans $ do
         cfor xlen $ \load_idx -> do
-          doArrayRef at aa True $ Right load_idx
-          sallocLet lv (return ()) $ do
-            store_let iv True load_idx $ do
-              cp (ca ra) body
+          let finalK = cp (ca ra) body
+          let bodyF (x, a) k () = do
+               doArrayRef at a True $ Right load_idx
+               sallocLet x (return ()) $
+                 store_let iv True load_idx $
+                 k
+          (F.foldrM bodyF finalK $ zip xs as) ()
```

Comment:
You definitely don't need these `()`s... what error do you get? (You do need to remove both of them)

Reply:
Removing them gives me this error:

```
/home/wgh/s/mk/reach-lang/hs/src/Reach/Connector/ALGO.hs:1942:19: error:
    • Couldn't match type ‘()’ with ‘ReaderT Env IO ()’
      Expected type: (DLVar, DLArg) -> App () -> ReaderT Env IO (App ())
        Actual type: (DLVar, DLArg)
                     -> App (App ()) -> ReaderT Env IO (App ())
    • In the first argument of ‘F.foldrM’, namely ‘bodyF’
      In the expression: F.foldrM bodyF finalK
      In a stmt of a 'do' block: (F.foldrM bodyF finalK $ zip xs as)
     |
1942 |         (F.foldrM bodyF finalK $ zip xs as)
     |                   ^^^^^
```

Maybe there's an obvious next step, but I'm not sure what it is.